### PR TITLE
fixed grad function of sinc

### LINKED ISSUE #53989: grad of sinc explodes at 0
`x = torch.tensor(0.0, requires_grad=True)`
`torch.autograd.grad(torch.sinc(x),x)`
gives:
`(tensor(nan),)`
and i see that in https://github.com/pytorch/pytorch/blob/master/tools/autograd/derivatives.yaml:
`grad * ((M_PI * self * (M_PI * self).cos() - (M_PI * self).sin()) / (M_PI * self * self)).conj()`
so it seems that 0 is not handled for the derivative??

## Environment

Collecting environment information...
PyTorch version: 1.8.0+cu101
Is debug build: False
CUDA used to build PyTorch: 10.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: 6.0.0-1ubuntu2 (tags/RELEASE_600/final)
CMake version: version 3.12.0

Python version: 3.7 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 11.0.221
GPU models and configuration: GPU 0: Tesla P100-PCIE-16GB
Nvidia driver version: 460.32.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.4
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.4
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.5
[pip3] torch==1.8.0+cu101
[pip3] torchsummary==1.5.1
[pip3] torchtext==0.9.0
[pip3] torchvision==0.9.0+cu101
[conda] Could not collect

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer @jianyuh @mruberry @heitorschueroff @walterddr @IvanYashchuk

### PR DESCRIPTION
**Pull request by johnpeterson123**
Summary:
Fixes #53989
Adds a special case for the sinc gradient formula to account for the case of input = 0

### CODE REVIEW

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 1]
```diff
@@ -1970,6 +1970,17 @@ Tensor elu_double_backward(
...
 
+Tensor sinc_backward(const Tensor& grad, const Tensor& self) {
+  Tensor self_scaled = at::ones_like(self);
+  Tensor scale_v;
+  Tensor pi = wrapped_scalar_tensor(Scalar(M_PI));
+  scale_v = grad *
+      ((pi * self * (pi * self).cos() - (pi * self).sin()) / (pi * self * self))
+          .conj();
+  scale_v.masked_fill_(self == 0, 0);
+  return self_scaled * scale_v;
```
Comment (soulitzer) [inline torch/csrc/autograd/FunctionsManual.cpp:1981]: Isn't `scale_v` already the size of `self`, why do we need to multiply by one here?

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 2]
```diff
@@ -1970,6 +1970,17 @@ Tensor elu_double_backward(
...
 }
 
+Tensor sinc_backward(const Tensor& grad, const Tensor& self) {
+  Tensor self_scaled = at::ones_like(self);
+  Tensor scale_v;
+  Tensor pi = wrapped_scalar_tensor(Scalar(M_PI));
+  scale_v = grad *
+      ((pi * self * (pi * self).cos() - (pi * self).sin()) / (pi * self * self))
+          .conj();
+  scale_v.masked_fill_(self == 0, 0);
```
Comment (soulitzer) [inline torch/csrc/autograd/FunctionsManual.cpp:1980]: From the failing test, looks like complex float/double isn't supported for masked_fill on CUDA. If we really want to unblock this, we'll have to drop support for complex on CUDA until that is supported for masked_fill, and we'll need to update the OpInfo to no longer test complex on CUDA. This is bc breaking though cc @gchanan

Comment (soulitzer) [reply torch/csrc/autograd/FunctionsManual.cpp:1980]: https://github.com/pytorch/pytorch/blob/aa8714dfedc73c67524e2394fe04d115f0783a09/torch/testing/_internal/common_methods_invocations.py

**Review by soulitzer — COMMENTED** [review]
Thanks for the fix!

Looks good overall, but it would be nice if you could add a small test for this case in `https://github.com/pytorch/pytorch/blob/master/test/test_unary_ufuncs.py`

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 3]
```diff
@@ -1970,6 +1970,17 @@ Tensor elu_double_backward(
     }
 }
 
+Tensor sinc_backward(const Tensor& grad, const Tensor& self) {
+  Tensor self_scaled = at::ones_like(self);
+  Tensor scale_v;
+  Tensor pi = wrapped_scalar_tensor(Scalar(M_PI));
+  scale_v = grad *
+      ((pi * self * (pi * self).cos() - (pi * self).sin()) / (pi * self * self))
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1978]: now that you moved this to an out of place function, you can pre-compute `pi * self` on the line above and re-use it here. That will improve speed quite a bit.

File: torch/csrc/autograd/FunctionsManual.cpp (hunk 2 above)
Comment (albanD) [reply torch/csrc/autograd/FunctionsManual.cpp:1980]: You can try to use where here that supports complex dtypes afaik
The python version would be this:
```
torch.where(mask, a, torch.tensor(0, dtype=a.dtype, device=a.device))
```

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 4]
```diff
@@ -1970,6 +1970,17 @@ Tensor elu_double_backward(
     }
 }
 
+Tensor sinc_backward(const Tensor& grad, const Tensor& self) {
+  Tensor self_scaled = at::ones_like(self);
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1974]: You don't use this below, you can just remove it.

**Review by albanD — COMMENTED** [review]
Thanks for sending a PR!

Added small comments below.

**Comment by johnpeterson123** [comment]
@soulitzer @albanD Thank you both for the review. I will make the changes that are discussed + add a test case then push another commit. Is it alright if at that time I ping you both for another review?

**Comment by albanD** [comment]
Hey,
Yes sure!
You can actually click the wheel next to the name of each reviewer in the list to ask for a new review!

**Comment by johnpeterson123** [comment]
@albanD `at::where()` is causing a runtime error "Expected scalar type float but found double. Is there a reason to opt for this over `at::nan_to_num()` which seems to be working properly? I just wanted to check before making another commit. All of the other comments have been dealt with without error.

**Comment by albanD** [comment]
The thing is that `nan_to_num` might be adding unrelated nans. Which is undesirable.

Is the error you see with where because the 0 you provide does not have the right type?

**Comment by johnpeterson123** [comment]
@albanD Yes it is a type error on the 0. I have tried casting it a few different ways. Do you have any hints? If not I will continue to mess with it until I can produce a desirable result.

**Comment by soulitzer** [comment]
@johnpeterson123 `torch.where` is pretty tricky when you pass a scalar:
> Currently valid scalar and tensor combination are 1. Scalar of floating dtype and torch.double 2. Scalar of integral dtype and torch.long 3. Scalar of complex dtype and torch.complex128

Have you tried passing the 0 as a tensor? Maybe something like this would work:
```
auto zeros = at::zeros_like(self);
auto result = at::where(self == 0, zeros, actual_values);
```

**Comment by johnpeterson123** [comment]
@albanD I had some more time to try to look into this. Looks like even using at::where there isn't support for torch.complex32 which might be where the failed tests are coming from? What do you think would be the best option to proceed?

File: test/test_unary_ufuncs.py [hunk 5]
```diff
@@ -1020,6 +1020,17 @@ def test_hardsigmoid_backward(self, device, dtype):
         out.backward(torch.ones_like(inputTensor))
         self.assertEqual(inputTensor.grad, expetedTensor)
 
+    @dtypesIfCUDA(torch.float, torch.double, torch.bfloat16)
```
Comment (albanD) [inline test/test_unary_ufuncs.py:1023]: From the test failure, you need to remove `bfloat16` from here as it is not supported.

File: torch/csrc/autograd/FunctionsManual.cpp [hunk 6]
```diff
@@ -1970,6 +1970,14 @@ Tensor elu_double_backward(
     }
 }
 
+Tensor sinc_backward(const Tensor& grad, const Tensor& self) {
+  Tensor pi_self = wrapped_scalar_tensor(Scalar(M_PI)) * self;
+  Tensor vals = grad *
+      ((pi_self * pi_self.cos() - pi_self.sin()) / (pi_self * self)).conj();
+  Tensor out = at::zeros_like(self, self.options());
```
Comment (albanD) [inline torch/csrc/autograd/FunctionsManual.cpp:1977]: I still don't think we want to do that.
You can see above in this file how to pass a scalar Tensor to where to avoid creating such a big Tensor full of 0s: https://github.com/pytorch/pytorch/blob/7c15862cc5ffddf9403c53f9a4b667f6ab331586/torch/csrc/autograd/FunctionsManual.cpp#L232

**Review by albanD — COMMENTED** [review]
Thanks for the update! 
Can you rebase on top of master as well to make sure all the CI can run properly?

**Comment by albanD** [comment]
Ho actually, I completely forgot that https://github.com/pytorch/pytorch/pull/56763 is already fixing this :)
